Modules/_abc.c

```diff
@@ -0,0 +1,960 @@
+        return NULL;
+    }
+    Py_SETREF(abc_invalidation_counter, next_version);
+
+    Py_INCREF(subclass);
+    return subclass;
+}
+
+PyDoc_STRVAR(_abc_instancecheck_doc,
+"Internal ABC helper for instance checks. Should be never used outside abc module");
+
+static PyObject *
+_abc_instancecheck(PyObject *m, PyObject *args)
+{
+    PyObject *self, *result = NULL, *subclass = NULL,
+             *subtype, *instance = NULL;
+    if (!PyArg_UnpackTuple(args, "_abc_instancecheck", 2, 2, &self, &instance)) {
+        return NULL;
+    }
+
+    _abc_data *impl = _get_impl(self);
+    if (!impl) {
+        return NULL;
+    }
+
+    subclass = _PyObject_GetAttrId(instance, &PyId___class__);
+    /* Inline the cache checking. */
+    int incache = _in_cache(impl, subclass);
+    if (incache < 0) {
+        goto end;
+    }
+    if (incache > 0) {
+        result = Py_True;
+        Py_INCREF(result);
+        goto end;
+    }
+    subtype = (PyObject *)Py_TYPE(instance);
+    if (subtype == subclass) {
+        int r = PyObject_RichCompareBool(
+            _get_negative_cache_version(impl),
+            abc_invalidation_counter, Py_EQ);
+        assert(r >= 0);  // Both should be PyLong
+        if (r > 0) {
+            incache = _in_negative_cache(impl, subclass);
+            if (incache < 0) {
+                goto end;
+            }
+            if (incache) {
+                result = Py_False;
+                Py_INCREF(result);
+                goto end;
+            }
+        }
+        /* Fall back to the subclass check. */
+        result = PyObject_CallMethod(self, "__subclasscheck__", "O", subclass);
+        goto end;
+    }
+    result = PyObject_CallMethod(self, "__subclasscheck__", "O", subclass);
+    if (!result || result == Py_True) {
+        goto end;
+    }
+    Py_DECREF(result);
+    result = PyObject_CallMethod(self, "__subclasscheck__", "O", subtype);
+
+end:
+    Py_CLEAR(impl);
+    Py_CLEAR(subclass);
+    return result;
+}
+
+
+PyDoc_STRVAR(_abc_subclasscheck_doc,
+"Internal ABC helper for subclasss checks. Should be never used outside abc module");
+
+static PyObject *
+_abc_subclasscheck(PyObject *m, PyObject *args)
+{
+    PyObject *self, *subclasses = NULL, *subclass = NULL, *result = NULL;
+    PyObject *ok, *mro, *key, *rkey;
+    Py_ssize_t pos;
+    int incache;
+    if (!PyArg_UnpackTuple(args, "_abc_subclasscheck", 2, 2, &self, &subclass)) {
+        return NULL;
+    }
+
+    _abc_data *impl = _get_impl(self);
+    if (!impl) {
+        return NULL;
+    }
+
+    /* 1. Check cache. */
+    incache = _in_cache(impl, subclass);
+    if (incache < 0) {
+        goto end;
+    }
+    if (incache > 0) {
+        result = Py_True;
+        goto end;
+    }
+
+    /* 2. Check negative cache; may have to invalidate. */
+    incache = _in_negative_cache(impl, subclass);
+    if (incache < 0) {
+        goto end;
+    }
+    if (PyObject_RichCompareBool(_get_negative_cache_version(impl),
```

Comment:
Check result of `_get_negative_cache_version` before passing it to `PyObject_RichCompareBool`

Reply:
Maybe if this is just a one line function that never fails we don't need it at all? Why not inline the field access?

Reply:
Exactly, this what I am doing now. Btw, allmost all of your new comments are already implemented, I just forgot to push, sorry for this! I will fix the remaining few and push all changes.

Reply:
NP, but please say when it's ready for the review again.